compiler/src/checker/checker.rs

```diff
@@ -6,11 +6,20 @@ use whistle_ast::IdentType;
 use whistle_ast::Literal;
 use whistle_ast::Operator;
 use whistle_ast::Primitive;
+use whistle_ast::Type;
+use whistle_common::Range;
+
+#[derive(Debug, Clone)]
+pub struct Constraint {
+  pub type1: Type,
+  pub type2: Type,
+  pub range: Option<Range>,
+}
 
 pub struct Checker {
   pub scope: ScopeContainer,
-  pub substitutions: Vec<IdentType>,
-  pub constraints: Vec<(IdentType, IdentType)>,
+  pub substitutions: Vec<Type>,
+  pub constraints: Vec<Constraint>,
   pub errors: Vec<CompilerError>,
```

Comment:
I think a generic and more extendable solution which we could use for the checker, lexer, parser and compiler would be a custom `DiagnosticHandler` struct which essentially is just a `Vec<Diagnostic>` where `Diagnostic` is something along the lines of:
```rs
enum Diagnostic {
  warning(Warning),
  error(Error)
}
```

That way we wouldn't have to dupe any error printing or handling code like what we have right now with the `throw` methods on the checker and compiler.

Reply:
Resolved by 536acff5dde53dfc5fe80af859009e84b9ab5156